### PR TITLE
Rhaas/scr prefix set

### PR DESCRIPTION
**Pull request by rhaas80**
This pull request adds functionality to let users set `SCR_PREFIX` either via `SCR_Config` or in their user config file if the `SCR_CONF_FILE` env. variable is set.

Changes are in the C code as well as some of the Perl scripts to read in `SCR_CONF_FILE` to be able to set `SCR_PREFIX` in there.

On the technical side this meant that I removed the reference counting and teardown of the various `XXX_hash` kvtrees that were being constantly created and torn down by `SCR_Config`. Right now this means that the first call to `scr_param_init` will initialize these structures from files ondisk but also that the first call to `scr_param_finalize` will tear them down.

`SCR_PREFIX` is special in that it determines the location of the file `app.conf` where `SCR_Config` values are saved in between `srun` calls so it uses custom parsing and it will be best of `SCR_PREFIX` is the first parameter set by a code.

### CODE REVIEW

**Comment by adammoody** [comment]
Oh, I meant to bring this up on the call, @rhaas80 .  I think it'd be nice to settle your work here as part of the next release.  It's been a while since we talked about it.  We could set aside time on the next group call to go over it.  I see it's still marked as draft, but I think it's close, right?

File: src/scr_param.c [hunk 1]
```diff
@@ -409,45 +406,63 @@ int scr_param_init()
...
 
+    /* initialize our hash for values from system config file */
+    scr_system_hash = kvtree_new();
+
+    /* safe to call scr_param_get after this */
+    scr_param_initialized = 1;
+
+    /* names we cannot read from the environment */
+    kvtree_set(scr_no_user_hash, "SCR_CNTL_BASE", kvtree_new());
+
```
Comment (adammoody) [inline src/scr_param.c:443]: We've switched to allow this now.
```suggestion
```

File: src/scr_param.c [hunk 2]
```diff
@@ -547,3 +558,16 @@ kvtree* scr_param_set_hash(char* name, kvtree* hash_value)
 
   return kvtree_set(scr_app_hash, name, hash_value);
 }
+
+/* unsets a parameter, returning 0 if the parameter did not exist */
+int scr_param_unset(char* name)
```
Comment (adammoody) [inline src/scr_param.c:563]: Can we upgrade this to a ``const char* name``?

File: src/scr_param.c [hunk 3]
```diff
@@ -547,3 +558,16 @@ kvtree* scr_param_set_hash(char* name, kvtree* hash_value)
...
+{
+  /* cannot set parameters that are used by scripts */
+  if (kvtree_get(scr_no_app_hash, name)) {
+    scr_dbg(1, "%s should not be changed at runtime if also using SCR scripts",
+      name
+    );
+  }
+
+  return kvtree_unset(scr_app_hash, name) == KVTREE_SUCCESS ? SCR_SUCCESS : SCR_FAILURE;
```
Comment (adammoody) [inline src/scr_param.c:572]: Let's split this in two which helps with running under a debugger.
```suggestion

  int rc = kvtree_unset(scr_app_hash, name);
  return (rc == KVTREE_SUCCESS) ? SCR_SUCCESS : SCR_FAILURE;
```

**Comment by adammoody** [comment]
Just a couple minor things, but this looks good.  Thanks, @rhaas80 .

File: src/scr_param.c (hunk 2 above)
Comment (rhaas80) [reply src/scr_param.c]: Done in 28dc2e7 where also `scr_param_set` and similar functions were modified to mark their `name` argument as const (no warnings were produced so they seem to pass them along to const functions on kvtree only).

File: examples/test_config.c [hunk 4]
```diff
@@ -197,46 +189,84 @@ int main (int argc, char* argv[])
...
+    scr_param_finalize();
+    SCR_Config("SCR_CONF_FILE="); /* undo setting SCR_CONF_FILE */
+    scr_param_finalize();
+
+    /* from env overriding user config file */
+    setenv("SCR_CONF_FILE", usrcfgfn, 1);
+    SCR_Config("SCR_PREFIX=.");
+    tests_passed &= test_cfg("SCR_COPY_TYPE", NULL, __LINE__);
+    scr_param_finalize();
+    SCR_Config("SCR_CONF_FILE="); /* undo setting SCR_PREFIX */
```
Comment (adammoody) [inline examples/test_config.c:238]: Do you mean to unset SCR_PREFIX here?

File: scripts/common/scr_param.pm.in [hunk 5]
```diff
@@ -13,64 +13,51 @@ sub new
...
-  
-  # tack on current working dir if needed
-  # don't resolve symlinks
-  # don't worry about missing parts, the calling script calling might create it
-  $prefix = `realpath -m -s $prefix`;
-  chomp $prefix;
+  # set our restricted parameters,
+  # these can't be set via env vars or user conf file
+  $self->{no_user} = {};
+  $self->{no_user}{"SCR_CNTL_BASE"} = {};
```
Comment (adammoody) [inline scripts/common/scr_param.pm.in:30]: We've switched SCR_CNTL_BASE to allow users to set it now.

**Comment by adammoody** [comment]
Thanks, @rhaas80 .  Your code here looks good.  I managed to introduce a merge conflict with a fix for the STORE test case.  I can help fix that up.

In the bigger picture, I'm still racking my brain on this one.  We'll want to modify things a bit to help the common case.  The good news is that people are using SCR_Config in *all* apps I have worked with.  So this is very popular.  In fact, the most common parameter that people set is ``SCR_PREFIX``.  Different apps set different sets of SCR parameters, but ``SCR_PREFIX`` is common across all of them.

They typically set ``SCR_PREFIX`` immediately at the point where they process an ``--output <dir>`` option that their user passes on the application command line to tell the app where it should writes its output files.  It's most natural for them to then immediately call ``SCR_Configf("SCR_PREFIX=%s", dir)``.  They similarly set other SCR parameters immediately upon processing corresponding application options.

The hard bit we run into is that the apps don't enforce an ordering on the command line options.  If even if we have to get a bit hacky, I'm trying to find a way so we don't require apps developers to either force an order on their users or force them to buffer up the SCR parameters along the way.  The first option is not user-friendly, and I've seen examples where the second option would be difficult.

My current thought is that we could allow someone to change ``SCR_PREFIX`` at any time, but with some well-defined consequences.  If SCR_PREFIX is changed, we could drop the kvtrees that correspond to files that are read from there (app.conf and user config) and re-read those files, i.e.,
```
SCR_Config("SCR_DEBUG=1")
  --> first call to SCR_Config
  --> in scr_param_init
  --> initialize kvtrees, init user config kvtree from current SCR_PREFIX value
SCR_Config("SCR_PREFIX=/new/dir")
  --> user has changed SCR_PREFIX
  --> drop user config file kvtree
  --> reinit user config file kvtree from file at /new/dir if one exists
```
The kvtree for the ``app.conf`` file is a bit harder.  I think to go down that road, we may need to split the app.conf kvtree into two different kvtrees.  One would hold the current app.conf file contents and the other would store values set by the user in the current run.  When SCR_PREFIX is changed, we can reinit the app.conf file kvtree but keep the other one unchanged.

On a query, we'd look to the current run kvtree first, then the app.conf file kvtree.  To write the app.conf file, we start with the app.conf kvtree and overwrite any keys with new values from the current run kvtree.

File: scripts/common/scr_param.pm.in (hunk 5 above)
Comment (rhaas80) [reply scripts/common/scr_param.pm.in]: Fixed in d4b8fac

File: examples/test_config.c (hunk 4 above)
Comment (rhaas80) [reply examples/test_config.c]: Fixed in 8f4cc25.

**Comment by rhaas80** [comment]
Ok, that should be doable.

Some more considereations when setting `SCR_PREFIX`:

1. `SCR_CONF_FILE` can be changed and `SCR_PREFIX` can be set in it, so one also has to re-initialize when `SCR_CONF_FILE` is set. 

2. splitting `app.conf` into a "current run" and "previous runs" parts requires that one merges the "current run" into the "previous runs" one at one point, the place to do that would seem `SCR_Init` which is the point where all settings go live and cannot be changed anymore anyway. Changing `SCR_PREFIX` requires that one removes the "current run" file from the previous location so that it is not read in by a `scr_param_init()` call in new run (reading the "current run" file seems required to support the case where there are `scr_param_init`...`scr_param_finalize`...`scr_param_init` sections of the code).

3. having multiple files requires extra care when removing of configuration values since it needs to interact well with the "unsetting reveals lower level parameters" mechanics desired among app.conf and the system configuration files. Namely one must actively record deletion of entries in the "current run" file so that `SCR_Config("FOO=")` records that `FOO` was removed and does not (only) remove it from the "current run" file (which would reveal its value for the `app.conf` file).

**Comment by adammoody** [comment]
@rhaas80 , in working with some recent apps, we're thinking we may want to avoid reading ``app.conf`` from SCR and moving to a mode where we just keep the keys in memory and still write the file at the end (so the scripts can still see the values used).  We can chat about it in a future meeting, but I wanted to give you a heads up since I know you're busy in this part of the code.  It may simplify things for you, as well.  Have a good weekend!

**Review by adammoody — APPROVED** [review]

**Comment by adammoody** [comment]
Thanks, @rhaas80 .  Got the changes here merged with https://github.com/LLNL/scr/pull/424.

**Pull request closed without merging**